mutation_writer/shard_based_splitting_writer.hh

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/util/noncopyable_function.hh>
+
+#include "feed_writers.hh"
+
+namespace mutation_writer {
+
+// Given a producer that may contain data for all shards, consume it in a per-shard
+// manner. This is useful, for instance, in the resharding process where a user changes
+// the amount of CPU assigned to Scylla and we have to rewrite the SSTables to their new
+// owners.
+future<> segregate_by_shard(flat_mutation_reader producer, reader_consumer consumer);
```

Comment:
How is the user expected to know which shard they're processing? with the other thing, they call this_shard_id() and know, but here it's all in the same shard.

Reply:
What is "the other thing"?

I am not in your head, Avi.

Shard information is derived from the partition keys that the producer produces.
I will add a comment about that.

Reply:
the other thing == distribute_reader_and_consume_on_shards()

I think it's better to make it explicit and pass the shard id as a parameter, rather than require the reader to inspect partition keys.

Isn't the shard id needed to pick the right generation when resharding?

Reply:
There is no need to know the shard_id to pick a generator, because resharding takes an std::function that creates the SSTable. The generation of the generation (wow) is encoded in that function.